### ISSUE TITLE
Add support for optimizer hints

### ISSUE THREAD

**Issue by Arivijay** [cockroachdb/activerecord-cockroachdb-adapter#266]
**Is your feature request related to a problem? Please describe.**
The CRDB Rails adapter is not currently equipped to process the different format being used by the customer for hinting no_zigzag_join/force_zig_zag. Customer wants to minimize the amount of effort required to migrate from Aurora to CRDB.

**Describe the solution you'd like**
Integrate the PR created by the customer into our CRDB Rails adapter. This involves being able to force an index on a query through optimizer hints
https://www.cockroachlabs.com/docs/v22.2/table-expressions#force-index-selection

**Describe alternatives you've considered**
The customer is manually incorporating RAW SQL queries into the ActiveRecord adapter which is a labor-intensive

Current ticket open 
https://github.com/cockroachlabs/support/issues/2155#issuecomment-1485671733

Jira issue: CRDB-26266

**Comment by mowinslow2** [comment]
[Here is a monkey patch to accomplish this task](https://gist.github.com/tkling/3976e6ae08e09fd2a7e005c558a85701), but will need to find a home in the adapter code for permanent resolution

**Comment by BuonOmo** [comment]
@Arivijay note that another alternative than raw sql is to use the `from` method, as done in the monkeypatch below.

@mowinslow2 thanks for the monkeypatch, I'm looking right now where it would be best to put it.

After looking a bit around in rails codebase, it is quite clear that the `connection_handler` has no effect on methods available for a `model`. E.g, the line below has no class related to connection:

```ruby
class Foo < ActiveRecord::Base; end
Foo.from("bar").ancestors
```

Which actually make sense since the connection can be changed dynamically. Another way to verify it is with multiple db connections:

```ruby
mgroup = %i(pg crdb mysql sqlite).map do |db|
  ApplicationRecord.connects_to(database: { writing: db, reading: db})
  Foo.connection
  Foo.from("bar").methods
end

mgroup.map.with_index do |ms, i|
  ms - [mgroup[0...i], mgroup[(i+1)..]].flatten
end
# => [[], [], [], []]
```

With all that say, this means that if we want the `force_index` and `index_hint` methods, we MUST monkeypatch an `ActiveRecord::Relation` related class.

However, what could be done is to have these methods depend on connection configuration.

1. Add a `support_index_hint?` to the `AbstractAdapter`
2. Override it in `CockroachAdapter`
3. choose a behaviour to run if not supported

For the behaviour i'd suggest to have this method somehow bubble up to a method in the connection adapter and then either run if in crdb, or raise `NotImplementedError` otherwise. That is what is done with the `ActiveRecord::Relation#explain` method, backed by `supports_explain?`.

WDYT?

**Comment by rafiss** [comment]
> Add a support_index_hint? to the AbstractAdapter

Just so I understand, this would require an upstream change in https://github.com/rails/rails, correct?

I think it would be preferable if the solution did not require code changes in any other project.

> However, what could be done is to have these methods depend on connection configuration.

Could you explain more about why we want to make the methods depend on connection configuration?

**Comment by BuonOmo** [comment]
No upstream push no. I don't believe this would benefit the whole ruby community as I don't know of any other connection adapter that would benefit from it.

Bubbling up to the connection adapter allows for someone using multiple database technologies to have a clear and distinct behaviour (either raising or doing nothing as mentioned in the end of my previous message, we'd have to choose).

So for instance:

```ruby
# This would not force index, and maybe either warn or even raise to tell user about the issue
PGUser.force_index("surname").first(10)

# This would just work
CRDBUser.force_index("surname").first(10)

**Comment by BuonOmo** [comment]
No upstream push no. I don't believe this would benefit the whole ruby community as I don't know of any other connection adapter that would benefit from it.

Bubbling up to the connection adapter allows for someone using multiple database technologies to have a clear and distinct behaviour (either raising or doing nothing as mentioned in the end of my previous message, we'd have to choose).

So for instance:

```ruby
# This would not force index, and maybe either warn or even raise to tell user about the issue
PGUser.force_index("surname").first(10)

# This would just work
CRDBUser.force_index("surname").first(10)
```

Also I just figured out that there is another case that I forgot about:

```ruby
User.force_index("name").from("special_user_table")
```

But I think this wouldn't be too hard to handle overall. And I suggest that the behaviour be : override forcing index since the table maybe completely different.

The example is a bit dull, but both the `force_index` and `from` could be in totally different files.

**Comment by rafiss** [comment]
> Bubbling up to the connection adapter allows for someone using multiple database technologies to have a clear and distinct behaviour (either raising or doing nothing as mentioned in the end of my previous message, we'd have to choose).

If someone is using this library, then we already know they are using CockroachDB, and not some other database technology.

---

I think your proposal seems fine. I am not understanding all the details, but I think the best way forward is to put your idea in a PR, and we can review it better there.

**Comment by BuonOmo** [comment]
> If someone is using this library, then we already know they are using CockroachDB, and not some other database technology.

[Yes, they can be!](https://github.com/cockroachdb/activerecord-cockroachdb-adapter/issues/251).

> I think your proposal seems fine. I am not understanding all the details, but I think the best way forward is to put your idea in a PR, and we can review it better there.

For sure, I just wanted to see if there was a big no go beforehand.

**Comment by BuonOmo** [comment]
There are some really tricky cases. In term of implementation for sure, but mostly in term of API design.

What do we want from both these requests?

```
User.force_index(...).from(...).index_hint(...)
User.force_index(...).index_hint(...).from(...)
```

There are multiple issues here. Mostly in terms of who is having the priority, and what should we keep in memory.

The [`ActiveRecord::QueryMethods#from`](https://devdocs.io/rails~7.0/activerecord/querymethods#method-i-from) methods can take either a table name, a list of table names, or a subquery. I  think that when it does take a subquery or a list of table names, we should ignore any previous or later call to either `force_index` or `index_hint`.

```ruby
# Just drop hint support, before or after
User.force_index(...).from(User.verified).index_hint(...)

# Keep hint support
User.force_index(...).from("other_table").index_hint(...)
```

However, in the second case, where we keep hint support, should we consider the `force_index`, that was defined beforehand? And what table should we apply the hint on?

My two cents here would be to say that — as opposed to the monkeypatch — we should always consider the last table mentionned in a `from` clause. So we should give priority to `#from` calls over index hint related calls.

In that same direction, I think we should also ignore hints that were set before a `#from` call.

**TL;DR:** `#from` should always have the priority over table optimizer hints. And hints should not persist along the query chain if `#from` is ever to be called.

That's what I'll take into account along my development. Feel free to ping me if you feel against it
